### DISCUSSION TITLE
Move the Device Flow Support behind a dedicated feature flag

### DISCUSSION THREAD

**Discussion by thomasdarimont** [keycloak/keycloak#23700] [Ideas]
The OAuth Device Flow capability has been shipped with Keycloak 13.0 and has since been enabled by default.
To use the Device Flow, one needs to create a client with the device flow enabled. Only with such a client device flow can it be used.
However, the device flow capability is always advertised via the [./well-known/openid-configuration](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/OIDCWellKnownProvider.java#L130) document.

Although Keycloak supports ways to customize the response of the openid-configuration endpoint, the endpoints to handle device-flow authentication are still exposed. Of course, those endpoints can be explicitly blocked with an external reverse proxy.

In most of our customer projects, we learned that customers would prefer to turn off device-flow support if they don't want to support it explicitly. Otherwise it gives their (3rd party) clients the wrong impression that certain functionality is supported when it is in fact not.

Therefore, I propose to put the support for OAuth2 Device Flow behind a dedicated feature flag like (DEVICE_FLOW). This flag could be enabled by default for a few versions and then turned off to give users time to adapt.

In addition, in some environments, moving this feature flag from a global flag to a realm-level flag might make sense, e.g., for SaaS scenarios where a realm is used as the unit of multi-tenancy.

The DEVICE_FLOW feature flag should be considered in the following places:

- While rendering information in the OpenID Discovery Document (./well-known/openid-configuration)
- While rendering the setting Device Flow checkbox in the admin-ui
- While rendering the token screen in the admin-ui
- When the device flow token is requested
- When the client capability config is shown

What do you think about that?

**Comment by thomasdarimont** [comment]
After giving this idea a bit more thought, I still think it would be helpful to be able the device flow feature for the whole server. However, I also think that it would be helpful to be able to explicitly enable / disable features like device-flow per realm.

**Reply by Michito-Okai** [reply]
> After giving this idea a bit more thought, I still think it would be helpful to be able the device flow feature for the whole server. However, I also think that it would be helpful to be able to explicitly enable / disable features like device-flow per realm.

This process has 2 steps.
Step1 is to add the device flow feature flag.
To enable device flow without adding settings, we should set the device flow feature flag to Type.DEFAULT in step1.
Thus, by setting Disabling features, you are able to set device flow disable for whole server.

Step2 is to be able to set enable / disable device flow per realm.
We need to discuss about Step2 because setting items of realm increase.
Also, we need to be able to set enable / disable not only device flow but also other features such as CIBA per realm.

Therefore, first, I want to implement stpe1.
What do you think?

**Reply by thomasdarimont** [reply]
Thanks for your comment. I think it is a good idea to do this incrementally. 

Step1 sounds reasonable to me. If we would implement a new `org.keycloak.common.Profile.Feature ` we can just add the appropriate checks around the device flow functionality and conditionally hide the configuration in the ui.

Also think Step2 (Realm-Level Feature Toggles) needs more planning as this will probably be a bigger feature, as we need a new way to handle features (Server-Level, Realm-Level) including admin ui extensions and feature toggle storage.

I'll create a draft PR for step1 and then we can go from there.

**Reply by thomasdarimont** [reply]
I created the issue https://github.com/keycloak/keycloak/issues/23891 for step1

**Reply by thomasdarimont** [reply]
@Michito-Okai I created a draft PR here: https://github.com/keycloak/keycloak/pull/23892

Would be great if you could take a look :)

I'll add a tests for the following Keycloak with the device_flow feature disabled:
- check that the device flow is no advertised via openid-configuration
- check that token request with device grant type produces an error
- check that device endpoints produce a not found error

**Reply by thomasdarimont** [reply]
I added some admin-ui screenshots with the device-flow feature disabled: https://github.com/keycloak/keycloak/pull/23892#issuecomment-1758344614
